File: Autonomous Mobile Robots Course/Chapter9/Lesson1/Chapter9_Lesson1.m
% Chapter 9 - Mapping Representations for Mobile Robots
% Lesson 1 - Occupancy Grid Mapping
%
% This script implements a simple 2D occupancy grid mapper using a log-odds grid.
% It also demonstrates a comparable workflow using Robotics System Toolbox (if available).
%
% Run:
%   Chapter9_Lesson1
%
% Notes:
% - For the from-scratch section, no toolboxes are required.
% - For the toolbox section, you need Robotics System Toolbox.

function Chapter9_Lesson1()
  close all; clc;

  % Map parameters
  W = 120; H = 100;
  res = 0.05;                % meters per cell
  origin = [0.0, 0.0];
  prior = 0.5;
  p_occ  = 0.75;
  p_free = 0.35;
  Lmin = -10; Lmax = 10;

  L0 = logit(prior);
  L  = L0 * ones(H, W);
  l_occ  = logit(p_occ)  - logit(prior);
  l_free = logit(p_free) - logit(prior);

  % Ground truth (only for generating synthetic measurements)
  gt = buildToyWorld(W, H);

  % Trajectory
  traj = zeros(25, 3);
  for k = 1:25
    traj(k,:) = [0.8 + 0.06*(k-1), 0.9 + 0.01*(k-1), 0.15];
  end

  % Sensor
  fov = pi;          % 180 deg
  nRays = 121;
  angles = linspace(-fov/2, fov/2, nRays);
  zMax = 3.0;

  % Mapping loop
  for k = 1:size(traj,1)
    pose = traj(k,:);
    for i = 1:numel(angles)
      ang = angles(i);
      z = castRayGrid(gt, origin, res, W, H, pose, ang, zMax);
      [L] = updateRay(L, origin, res, W, H, pose, ang, z, zMax, l_free, l_occ, Lmin, Lmax);
    end
  end

  % Visualize occupancy probabilities
  P = inv_logit(L);
  figure('Name','Occupancy probability (from scratch)');
  imagesc(P); axis image; set(gca, 'YDir','normal');
  title('Occupancy probability (after mapping)');
  xlabel('grid x'); ylabel('grid y'); colorbar;

  figure('Name','Ground truth (simulation only)');
  imagesc(gt); axis image; set(gca, 'YDir','normal');
  title('Ground truth occupancy (used only for synthetic rays)');
  xlabel('grid x'); ylabel('grid y');

  % Optional: Robotics System Toolbox demonstration
  demoRoboticsToolboxIfAvailable(gt, res);
end

function demoRoboticsToolboxIfAvailable(gt, res)
  if exist('occupancyMap','class') ~= 8
    fprintf('Robotics System Toolbox not detected; skipping toolbox demo.\n');
    return;
  end

  fprintf('Robotics System Toolbox detected; running a brief occupancyMap demo...\n');
  map = occupancyMap(size(gt,2)*res, size(gt,1)*res, 1/res); % width[m], height[m], cells/m
  setOccupancy(map, gt);  % interpret gt as occupancy (0/1)
  figure('Name','Toolbox: occupancyMap visualization');
  show(map); title('Toolbox map visualization');
end

function gt = buildToyWorld(W, H)
  gt = zeros(H, W, 'uint8');

  % Border walls
  gt(1,:)   = 1;
  gt(end,:) = 1;
  gt(:,1)   = 1;
  gt(:,end) = 1;

  % Rectangle obstacle
  gt(31:50, 56:75) = 1;

  % Small obstacle
  gt(71:80, 26:35) = 1;
end

function z = castRayGrid(gt, origin, res, W, H, pose, relAng, zMax)
  step = 0.5 * res;
  x = pose(1); y = pose(2); theta = pose(3);
  a = theta + relAng;

  dist = 0.0;
  while dist < zMax
    xt = x + dist*cos(a);
    yt = y + dist*sin(a);
    [gx, gy] = worldToGrid(origin, res, xt, yt);
    if gx >= 1 && gx <= W && gy >= 1 && gy <= H
      if gt(gy, gx) == 1
        z = dist; return;
      end
    else
      z = dist; return;
    end
    dist = dist + step;
  end
  z = zMax;
end

function [L] = updateRay(L, origin, res, W, H, pose, relAng, z, zMax, l_free, l_occ, Lmin, Lmax)
  x = pose(1); y = pose(2); theta = pose(3);
  z = max(0.0, min(z, zMax));
  a = theta + relAng;

  xe = x + z*cos(a);
  ye = y + z*sin(a);

  [x0, y0] = worldToGrid(origin, res, x, y);
  [x1, y1] = worldToGrid(origin, res, xe, ye);

  if x0 < 1 || x0 > W || y0 < 1 || y0 > H
    return;
  end

  cells = bresenham(x0, y0, x1, y1);
  if isempty(cells), return; end

  hit = (z < zMax - 1e-6);

  if hit
    freeCells = cells(1:end-1,:);
  else
    freeCells = cells;
  end

  for i = 1:size(freeCells,1)
    cx = freeCells(i,1); cy = freeCells(i,2);
    if cx >= 1 && cx <= W && cy >= 1 && cy <= H
      L(cy, cx) = min(Lmax, max(Lmin, L(cy, cx) + l_free));
    end
  end

  if hit
    cx = cells(end,1); cy = cells(end,2);
    if cx >= 1 && cx <= W && cy >= 1 && cy <= H
      L(cy, cx) = min(Lmax, max(Lmin, L(cy, cx) + l_occ));
    end
  end
end

function [gx, gy] = worldToGrid(origin, res, x, y)
  gx = floor((x - origin(1)) / res) + 1; % MATLAB 1-based
  gy = floor((y - origin(2)) / res) + 1;
end

function pts = bresenham(x0, y0, x1, y1)
  dx = abs(x1 - x0);
  dy = abs(y1 - y0);
  sx = 1; if x1 < x0, sx = -1; end
  sy = 1; if y1 < y0, sy = -1; end
  err = dx - dy;

  x = x0; y = y0;
  pts = zeros(0,2);
  while true
    pts(end+1,:) = [x, y]; %#ok<AGROW>
    if x == x1 && y == y1
      break;
    end
    e2 = 2*err;
    if e2 > -dy
      err = err - dy;
      x = x + sx;
    end
    if e2 < dx
      err = err + dx;
      y = y + sy;
    end
  end
end

function y = logit(p)
  eps = 1e-12;
  p = min(1.0-eps, max(eps, p));
  y = log(p/(1-p));
end

function p = inv_logit(L)
  p = 1 ./ (1 + exp(-L));
end

File: Autonomous_Mobile_Robots_Course/Chapter9/Lesson1/Chapter9_Lesson1.py
"""
Chapter 9 - Mapping Representations for Mobile Robots
Lesson 1 - Occupancy Grid Mapping

This script implements a simple 2D occupancy grid mapper using a log-odds grid and
a very basic inverse sensor model for a planar range sensor (e.g., 2D LiDAR).
It also includes a small synthetic simulation to generate range readings in a toy world.

Dependencies:
  - numpy
  - matplotlib

Run:
  python Chapter9_Lesson1.py
"""

from __future__ import annotations
import math
from dataclasses import dataclass
from typing import Iterable, List, Tuple

import numpy as np
import matplotlib.pyplot as plt


def logit(p: float) -> float:
    p = float(np.clip(p, 1e-9, 1.0 - 1e-9))
    return math.log(p / (1.0 - p))


def inv_logit(l: np.ndarray) -> np.ndarray:
    return 1.0 / (1.0 + np.exp(-l))


def bresenham(x0: int, y0: int, x1: int, y1: int) -> List[Tuple[int, int]]:
    """
    Integer grid traversal for a line segment (x0,y0)->(x1,y1).
    Returns list of (x,y) cells including both endpoints.
    """
    pts: List[Tuple[int, int]] = []
    dx = abs(x1 - x0)
    dy = abs(y1 - y0)
    sx = 1 if x1 >= x0 else -1
    sy = 1 if y1 >= y0 else -1
    err = dx - dy

    x, y = x0, y0
    while True:
        pts.append((x, y))
        if x == x1 and y == y1:
            break
        e2 = 2 * err
        if e2 > -dy:
            err -= dy
            x += sx
        if e2 < dx:
            err += dx
            y += sy
    return pts


@dataclass
class OccupancyGrid:
    width: int
    height: int
    resolution: float  # meters per cell
    origin: Tuple[float, float] = (0.0, 0.0)  # world coords of grid (0,0) cell corner
    prior: float = 0.5
    p_occ: float = 0.7
    p_free: float = 0.3
    l_min: float = -10.0
    l_max: float = 10.0

    def __post_init__(self) -> None:
        self.L = np.full((self.height, self.width), logit(self.prior), dtype=np.float64)
        # Log-odds increments (relative to prior odds)
        self.l_occ = logit(self.p_occ) - logit(self.prior)
        self.l_free = logit(self.p_free) - logit(self.prior)

    def world_to_grid(self, x: float, y: float) -> Tuple[int, int]:
        gx = int(math.floor((x - self.origin[0]) / self.resolution))
        gy = int(math.floor((y - self.origin[1]) / self.resolution))
        return gx, gy

    def in_bounds(self, gx: int, gy: int) -> bool:
        return 0 <= gx < self.width and 0 <= gy < self.height

    def update_ray(
        self,
        pose: Tuple[float, float, float],
        rel_angle: float,
        rng: float,
        max_range: float,
    ) -> None:
        """
        Update the grid with a single range measurement.
        pose = (x, y, theta) in world coordinates [m, m, rad]
        rel_angle is sensor ray angle relative to robot heading.
        rng is measured range in meters (clipped to max_range).
        """
        x, y, theta = pose
        rng = float(np.clip(rng, 0.0, max_range))
        a = theta + rel_angle

        # Ray end point in world coordinates
        xe = x + rng * math.cos(a)
        ye = y + rng * math.sin(a)

        x0, y0 = self.world_to_grid(x, y)
        x1, y1 = self.world_to_grid(xe, ye)

        if not self.in_bounds(x0, y0):
            return  # robot outside the map (ignore)
        # Traverse cells; all but last are "free". Last is "occupied" if hit before max_range.
        cells = bresenham(x0, y0, x1, y1)
        if len(cells) == 0:
            return

        hit = (rng < max_range - 1e-6)
        free_cells = cells[:-1] if hit else cells  # if no hit, entire ray is free (to max range)

        for cx, cy in free_cells:
            if self.in_bounds(cx, cy):
                self.L[cy, cx] = np.clip(self.L[cy, cx] + self.l_free, self.l_min, self.l_max)

        if hit:
            cx, cy = cells[-1]
            if self.in_bounds(cx, cy):
                self.L[cy, cx] = np.clip(self.L[cy, cx] + self.l_occ, self.l_min, self.l_max)

    def prob(self) -> np.ndarray:
        return inv_logit(self.L)


def build_toy_world(width: int, height: int) -> np.ndarray:
    """
    Ground truth occupancy for synthetic ray casting.
    0 = free, 1 = occupied
    """
    world = np.zeros((height, width), dtype=np.uint8)
    # Border walls
    world[0, :] = 1
    world[-1, :] = 1
    world[:, 0] = 1
    world[:, -1] = 1
    # A rectangle obstacle
    world[30:50, 55:75] = 1
    # A smaller obstacle
    world[70:80, 25:35] = 1
    return world


def cast_ray_grid(
    world_occ: np.ndarray,
    grid: OccupancyGrid,
    pose: Tuple[float, float, float],
    rel_angle: float,
    max_range: float,
    step: float | None = None,
) -> float:
    """
    Simple ray casting by stepping along the ray in world coords until an occupied cell is hit.
    Returns the measured range.
    """
    if step is None:
        step = grid.resolution * 0.5

    x, y, theta = pose
    a = theta + rel_angle
    dist = 0.0
    while dist < max_range:
        xt = x + dist * math.cos(a)
        yt = y + dist * math.sin(a)
        gx, gy = grid.world_to_grid(xt, yt)
        if 0 <= gx < grid.width and 0 <= gy < grid.height:
            if world_occ[gy, gx] == 1:
                return dist
        else:
            # out of bounds acts like a wall
            return dist
        dist += step
    return max_range


def main() -> None:
    # Map parameters
    W, H = 120, 100
    res = 0.05  # 5 cm per cell
    og = OccupancyGrid(W, H, res, origin=(0.0, 0.0), prior=0.5, p_occ=0.75, p_free=0.35)

    # Ground truth (for generating measurements only)
    gt = build_toy_world(W, H)

    # Robot trajectory (world meters)
    traj = []
    for k in range(25):
        x = 0.8 + 0.06 * k
        y = 0.9 + 0.01 * k
        theta = 0.15  # rad
        traj.append((x, y, theta))

    # Sensor model
    fov = math.radians(180.0)
    n_rays = 121
    angles = np.linspace(-fov / 2.0, fov / 2.0, n_rays)
    z_max = 3.0  # meters

    # Mapping loop
    for pose in traj:
        for ang in angles:
            z = cast_ray_grid(gt, og, pose, float(ang), z_max)
            og.update_ray(pose, float(ang), z, z_max)

    # Visualize: probability of occupancy
    P = og.prob()
    plt.figure(figsize=(7, 5))
    plt.title("Occupancy probability (after mapping)")
    plt.imshow(P, origin="lower", vmin=0.0, vmax=1.0)
    plt.xlabel("grid x")
    plt.ylabel("grid y")
    plt.colorbar(label="p(occupied)")
    plt.tight_layout()
    plt.show()

    # Also visualize ground truth for comparison (optional)
    plt.figure(figsize=(7, 5))
    plt.title("Ground truth occupancy (used only for simulation)")
    plt.imshow(gt, origin="lower", vmin=0, vmax=1)
    plt.xlabel("grid x")
    plt.ylabel("grid y")
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()
